### DISCUSSION TITLE
How to create/add a new address space in a Loader?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#6907] [Q&A]
I'm making some loaders for the '80s microcomputer the TI-99/4A which had an unusual architecture.

As well as the usual address space the CPU can access with ROM and RAM it also had a "GROM" address space which was graphics RAM and where byte code called GPL (graphics programming language) resided. I want to add a "GROM" address space.

I don't usually program in Java. I can find `protected` `addAddressSpace()` member function in `DefaultAddressFactory.java` but since it's protected I can't call it. The only code I can find which does use it is in `GadpClient.java` and `VarnodeContext.java`, which both `extend` `DefaultAddressFactory`

This is getting too far down the Java rabbit hole for me and my naïve attempts to implement it are failing. I can't find any Ghidra Loader which adds an additional address space either.

**Comment by ryanmkurtz** [comment]
Address spaces are defined by a processor, not a loader.  So, it will have to be added to your TI-99/4A processor module (.slaspec file) for Ghidra to know about it.

**Comment by hippietrail** [comment]
Ah I started thinking it could be something like that. This is an odd platform and the CPU can't access the GROM address space. It communicates via a couple of hardware registers.
I'm not sure if a better approach could be to define an abstract processor that wraps both the CPU and the GROM+GPL. Or maybe it would need a .slaspec for the TMS990 CPU and another for the GROM+GPL. I haven't looked at how loaders for platforms with multiple different processors work. Maybe it's simply not a good fit for Ghidra.
